apps/dash-oil-gas-ternary/app.py

```diff
@@ -0,0 +1,791 @@
+            for formation in formations:
+                if bar_click_data is None:
+                    processed_data[formation] = None
+                elif formation not in processed_data:
+                    processed_data[formation] = []
+
+        else:
+
+            for formation in formations:
+                processed_data[formation] = None
+
+        return generate_ternary_map(
+            dff, processed_data, contour_visible, marker_visible
+        )
+
+    if prop_id == "ternary-layer-select":
+        if curr_fig is not None:
+            if "Well Data" not in layer_select:
+                marker_visible = "legendonly"
+            if "Rock Type" not in layer_select:
+                contour_visible = "legendonly"
+
+            for contour_dict in curr_fig["data"][:32]:
+                contour_dict["visible"] = contour_visible
+
+            for marker_dict in curr_fig["data"][32:]:
+                marker_dict["visible"] = marker_visible
+            return curr_fig
+        else:
+            return curr_fig
+
+
+# Update well map
+@app.callback(
+    output=Output("well-map", "figure"),
+    inputs=[
+        Input("ternary-map", "selectedData"),
+        Input("form-by-bar", "selectedData"),
+        Input("form-by-bar", "clickData"),
+        Input("operator-select", "value"),
+        Input("mapbox-view-selector", "value"),
+    ],
+)
+def update_well_map(
+    tern_selected_data, bar_selected_data, bar_click_data, op_select, mapbox_view
+):
+    dff = df[df["op"].isin(op_select)]
+    formations = dff["fm_name"].unique().tolist()
+
+    # Find which one has been triggered
+    ctx = dash.callback_context
+
+    prop_id = ""
+    prop_type = ""
+    if ctx.triggered:
+        splitted = ctx.triggered[0]["prop_id"].split(".")
+        prop_id = splitted[0]
+        prop_type = splitted[1]
+
+    processed_data = {}
+
+    if prop_id == "ternary-map":
+        for formation in formations:
+            if tern_selected_data is None:
+                processed_data[formation] = None
+            else:
+                processed_data[formation] = get_selection(
+                    dff, formation, tern_selected_data, 32
+                )
+
+    elif prop_id == "form-by-bar":
+
+        bar_data = ""
+        if prop_type == "selectedData":
+            bar_data = bar_selected_data
+        elif prop_type == "clickData":
+            bar_data = bar_click_data
+
+        processed_data = get_selection_by_bar(bar_data)
+
+        for formation in formations:
+            if bar_data is None:
+                processed_data[formation] = None
+            elif formation not in processed_data:
+                processed_data[formation] = []
+
+    else:
+        for formation in formations:
+            processed_data[formation] = None
+
+    return generate_well_map(dff, processed_data, mapbox_view)
+
+
+# Update production plot
+@app.callback(
+    output=Output("production-fig", "figure"),
```

Comment:
We actually don't typically include these keywords in the arguments of a callback. 
```suggestion
    Output("production-fig", "figure"),
```